// @jshookmcp/extension-sdk@0.3.0/dist/plugin.js
import { toTextResponse, toErrorResponse, } from './bridges/shared.js';
// ── Response helpers (delegates to bridge) ──
/** Build a success JSON response for an MCP tool. Alias of `toTextResponse`. */
export const jsonResponse = toTextResponse;
/** Build an error JSON response for an MCP tool. Alias of `toErrorResponse`. */
export const errorResponse = toErrorResponse;
// ── ExtensionBuilder ──
export class ExtensionBuilder {
    // ── state ──
    _id;
    _version;
    _name = '';
    _description = '';
    _author = '';
    _sourceRepo = '';
    _compatibleCore = '>=0.1.0';
    _profiles = ['full'];
    _tools = [];
    _allowCommands = [];
    _allowHosts = [];
    _allowTools = [];
    _metrics = [];
    _configDefaults = {};
    _onLoadHandler;
    _onValidateHandler;
    _onActivateHandler;
    _onDeactivateHandler;
    constructor(id, version) {
        this._id = id;
        this._version = version;
    }
    // ── accessors ──
    get id() { return this._id; }
    get version() { return this._version; }
    get pluginName() { return this._name; }
    get pluginDescription() { return this._description; }
    get pluginAuthor() { return this._author; }
    get pluginSourceRepo() { return this._sourceRepo; }
    get compatibleCoreRange() { return this._compatibleCore; }
    get profiles() { return this._profiles; }
    get tools() { return this._tools; }
    get allowedCommands() { return this._allowCommands; }
    get allowedHosts() { return this._allowHosts; }
    get allowedTools() { return this._allowTools; }
    get declaredMetrics() { return this._metrics; }
    get configDefaults() { return this._configDefaults; }
    get onLoadHandler() { return this._onLoadHandler; }
    get onValidateHandler() { return this._onValidateHandler; }
    get onActivateHandler() { return this._onActivateHandler; }
    get onDeactivateHandler() { return this._onDeactivateHandler; }
    // ── setters ──
    name(n) { this._name = n; return this; }
    description(desc) { this._description = desc; return this; }
    author(a) { this._author = a; return this; }
    sourceRepo(url) { this._sourceRepo = url; return this; }
    compatibleCore(range) { this._compatibleCore = range; return this; }
    /**
     * Merge external metadata (e.g. from meta.yaml) into the builder.
     * Only fills in fields that the builder chain has NOT explicitly set
     * (i.e. still at their default empty-string value).
     */
    mergeMetadata(meta) {
        if (!this._name && meta.name)
            this._name = meta.name;
        if (!this._description && meta.description)
            this._description = meta.description;
        if (!this._author && meta.author)
            this._author = meta.author;
        if (!this._sourceRepo && meta.source_repo)
            this._sourceRepo = meta.source_repo;
        return this;
    }
    profile(p) {
        this._profiles = Array.isArray(p) ? p : [p];
        return this;
    }
    allowCommand(cmd) { this._allowCommands.push(...(Array.isArray(cmd) ? cmd : [cmd])); return this; }
    allowHost(host) { this._allowHosts.push(...(Array.isArray(host) ? host : [host])); return this; }
    allowTool(tool) { this._allowTools.push(...(Array.isArray(tool) ? tool : [tool])); return this; }
    metric(m) { this._metrics.push(...(Array.isArray(m) ? m : [m])); return this; }
    configDefault(key, value) { this._configDefaults[key] = value; return this; }
    /**
     * Register a tool exposed by this extension.
     *
     * @param name    Unique tool name (must not collide with built-in tools).
     * @param desc    Human-readable description shown to the AI model.
     * @param schema  JSON-Schema **properties** object — the builder automatically
     *                wraps it in `{ type: 'object', properties: … }`, so you only
     *                need to pass the inner properties map.
     *                Example: `{ text: { type: 'string', description: 'Input' } }`
     * @param handler Async function `(args, ctx) => ToolResponse`.
     * @param profiles Optional profile tiers for this specific tool (defaults to extension-level profiles).
     */
    tool(name, desc, schema, handler, profiles) {
        this._tools.push({
            name,
            description: desc,
            schema: { type: 'object', properties: schema },
            handler,
            profiles,
        });
        return this;
    }
    onLoad(h) { this._onLoadHandler = h; return this; }
    onValidate(h) { this._onValidateHandler = h; return this; }
    onActivate(h) { this._onActivateHandler = h; return this; }
    onDeactivate(h) { this._onDeactivateHandler = h; return this; }
}
export function createExtension(id, version) {
    return new ExtensionBuilder(id, version);
}
