// @jshookmcp/extension-sdk/dist/workflow.js
export class WorkflowNodeBuilder {
    id;
    constructor(id) {
        this.id = id;
    }
}
export class ToolNodeBuilder extends WorkflowNodeBuilder {
    toolName;
    _input;
    _retry;
    _timeoutMs;
    constructor(id, toolName) {
        super(id);
        this.toolName = toolName;
    }
    input(input) {
        this._input = input;
        return this;
    }
    retry(policy) {
        this._retry = policy;
        return this;
    }
    timeout(ms) {
        this._timeoutMs = ms;
        return this;
    }
    build() {
        return {
            kind: 'tool',
            id: this.id,
            toolName: this.toolName,
            input: this._input,
            retry: this._retry,
            timeoutMs: this._timeoutMs,
        };
    }
}
/**
 * Shared base for builders that contain child steps (Sequence and Parallel).
 * Eliminates duplicated `step()` / `tool()` / `sequence()` / `parallel()` / `branch()` methods.
 */
class CompositeNodeBuilder extends WorkflowNodeBuilder {
    _steps = [];
    step(nodeBuilder) {
        this._steps.push(nodeBuilder);
        return this;
    }
    /**
     * Add a tool node.
     *
     * Accepts either an options object for simple cases or a callback for full
     * control:
     *
     * ```ts
     * .tool('nav', 'page_navigate', { input: { url: '...' } })
     * .tool('nav', 'page_navigate', (b) => b.input({ url: '...' }).timeout(5000))
     * ```
     */
    tool(id, toolName, config) {
        const builder = new ToolNodeBuilder(id, toolName);
        if (config) {
            if (typeof config === 'function') {
                config(builder);
            }
            else {
                if (config.input)
                    builder.input(config.input);
                if (config.retry)
                    builder.retry(config.retry);
                if (config.timeoutMs !== undefined)
                    builder.timeout(config.timeoutMs);
            }
        }
        this._steps.push(builder);
        return this;
    }
    sequence(id, config) {
        const builder = new SequenceNodeBuilder(id);
        if (config)
            config(builder);
        this._steps.push(builder);
        return this;
    }
    parallel(id, config) {
        const builder = new ParallelNodeBuilder(id);
        if (config)
            config(builder);
        this._steps.push(builder);
        return this;
    }
    branch(id, predicateId, config) {
        const builder = new BranchNodeBuilder(id, predicateId);
        if (config)
            config(builder);
        this._steps.push(builder);
        return this;
    }
}
export class SequenceNodeBuilder extends CompositeNodeBuilder {
    build() {
        return {
            kind: 'sequence',
            id: this.id,
            steps: this._steps.map((b) => b.build()),
        };
    }
}
export class ParallelNodeBuilder extends CompositeNodeBuilder {
    _maxConcurrency = 4;
    _failFast = false;
    maxConcurrency(concurrency) {
        this._maxConcurrency = concurrency;
        return this;
    }
    failFast(ff) {
        this._failFast = ff;
        return this;
    }
    build() {
        return {
            kind: 'parallel',
            id: this.id,
            steps: this._steps.map((b) => b.build()),
            maxConcurrency: this._maxConcurrency,
            failFast: this._failFast,
        };
    }
}
export class BranchNodeBuilder extends WorkflowNodeBuilder {
    predicateId;
    _predicateFn;
    _whenTrue;
    _whenFalse;
    constructor(id, predicateId) {
        super(id);
        this.predicateId = predicateId;
    }
    predicateFn(fn) {
        this._predicateFn = fn;
        return this;
    }
    whenTrue(nodeBuilder) {
        this._whenTrue = nodeBuilder;
        return this;
    }
    whenFalse(nodeBuilder) {
        this._whenFalse = nodeBuilder;
        return this;
    }
    build() {
        if (!this._whenTrue) {
            throw new Error(`BranchNode '${this.id}' requires a whenTrue step`);
        }
        return {
            kind: 'branch',
            id: this.id,
            predicateId: this.predicateId,
            predicateFn: this._predicateFn,
            whenTrue: this._whenTrue.build(),
            whenFalse: this._whenFalse ? this._whenFalse.build() : undefined,
        };
    }
}
export class WorkflowBuilder {
    _id;
    _displayName;
    _description;
    _tags;
    _timeoutMs;
    _defaultMaxConcurrency;
    _buildFn;
    _onStart;
    _onFinish;
    _onError;
    constructor(id, displayName) {
        this._id = id;
        this._displayName = displayName;
    }
    description(desc) { this._description = desc; return this; }
    tags(tags) { this._tags = tags; return this; }
    timeoutMs(timeout) { this._timeoutMs = timeout; return this; }
    defaultMaxConcurrency(max) { this._defaultMaxConcurrency = max; return this; }
    buildGraph(fn) {
        this._buildFn = (ctx) => fn(ctx).build();
        return this;
    }
    onStart(fn) { this._onStart = fn; return this; }
    onFinish(fn) { this._onFinish = fn; return this; }
    onError(fn) { this._onError = fn; return this; }
    build() {
        if (!this._buildFn)
            throw new Error(`WorkflowBuilder '${this._id}' needs a buildGraph() function.`);
        return {
            kind: 'workflow-contract',
            version: 1,
            id: this._id,
            displayName: this._displayName,
            description: this._description,
            tags: this._tags,
            timeoutMs: this._timeoutMs,
            defaultMaxConcurrency: this._defaultMaxConcurrency,
            build: this._buildFn,
            onStart: this._onStart,
            onFinish: this._onFinish,
            onError: this._onError,
        };
    }
}
export function createWorkflow(id, displayName) {
    return new WorkflowBuilder(id, displayName);
}
// ── Convenience factory functions ──
/** Create a tool node. */
export function toolNode(id, toolName) {
    return new ToolNodeBuilder(id, toolName);
}
/** Create a sequence node. */
export function sequenceNode(id) {
    return new SequenceNodeBuilder(id);
}
/** Create a parallel node. */
export function parallelNode(id) {
    return new ParallelNodeBuilder(id);
}
/** Create a branch node. */
export function branchNode(id, predicateId) {
    return new BranchNodeBuilder(id, predicateId);
}
